### PR TITLE
compiler: Decouple pthreading pass from orchestration

### PR DESCRIPTION
**Pull request by FabioLuporini**
No new functionalities added. This PR rewrites some ugly old code and decouple a big pass into two passes. This paves the way to future passes requiring asynchronous tasking

### CODE REVIEW

File: devito/passes/iet/asynchrony.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
+import cgen as c
```
Comment (FabioLuporini) [inline devito/passes/iet/asynchrony.py:1]: note for reviewers: this whole new file stems from `ir/iet/efunc.py`

File: devito/types/parallel.py [hunk 2]
```diff
@@ -237,152 +240,6 @@ def _C_typedata(self):
     def locked_dimensions(self):
         return set().union(*[d._defines for d in self.dimensions])
 
-    # Pickling support
```
Comment (FabioLuporini) [inline devito/types/parallel.py:240]: note for reviewers: most of this code has been moved to `ir/support/syncs.py` (and partly rewritten)

File: tests/test_gpu_common.py [hunk 3]
```diff
@@ -588,57 +579,6 @@ def test_streaming_postponed_deletion(self, opt, ntmps):
...
-                Eq(u.forward, f + 1)]
-
-        op = Operator(eqns, opt=('streaming', 'orchestrate'))
-
-        assert len(op._func_table) == 3
-        assert 'init_device0' in op._func_table
-        assert 'prefetch_host_to_device0' in op._func_table
-
-    @skipif('device-openmp')  # TODO: Still unsupported with OpenMP, but soon will be
-    def test_composite_streaming_tasking(self):
```
Comment (FabioLuporini) [inline tests/test_gpu_common.py:607]: note for reviewers: moved below

File: devito/passes/clusters/asynchrony.py [hunk 4]
```diff
@@ -239,165 +322,3 @@ def is_memcpy(cluster):
...
-        pfetch, pcond,
-        target, tstore
-    ))
-    actions[last].insert.append(pcluster)
-    actions[cluster].drop = True
-
-    return last, pcluster
-
-
-def actions_from_unstructured(clusters, key, prefix, actions):
```
Comment (FabioLuporini) [inline devito/passes/clusters/asynchrony.py:359]: note for reviewers: dropped as legacy and useless

**Review by mloubout — APPROVED** [review]
Looks sane from wuick pass through, mostly moving stuff and cleanup

**Review by georgebisbas — APPROVED** [review]

**Pull request merged**